### ISSUE TITLE
[ELE-50] Support for MS Teams alerts

### ISSUE THREAD

**Issue by bchin3** [elementary-data/elementary#324]
It would be great to be able to receive alerts via MS Teams, as an alternative to Slack

<sub>[ELE-50](https://linear.app/elementary/issue/ELE-50/support-for-ms-teams-alerts)</sub>

**Comment by daniel-gomez-mng** [comment]
Hi there! I am interested in use elementary in our project and feature would be very important. Would it be possible to reopen the issue? Thank you!

**Comment by abi3050** [comment]
It’s the same for me. We want to use Elementary if it has the feature to notify Teams Channels. We’d also be happy to see more Microsoft integrations, i.e. support for hosting on Azure instead of just S3 buckets. Thanks!

**Comment by Januka-Peiris-Assured** [comment]
Hello

I would like to assist with this if possible. Just want to understand if the current notifications to Slack is sent via API or another method.

**Comment by oravi** [comment]
Hi @Januka-Peiris-Assured ! 

Here's an high-level overview on what's needed in order to implement this integration, please feel free to reach out here or over Slack if you have further questions. Happy to help and hop on a call if needed. 

- [monitor/cli.py](https://github.com/elementary-data/elementary/blob/master/elementary/monitor/cli.py#L169) -
    - This is the file where the CLI entry points and their parameters are defined. We need to add to the `monitor` command a new set of parameters to get the MS teams connection details (token / webhook).
- [config/config.py -](https://github.com/elementary-data/elementary/blob/master/elementary/config/config.py)
    - In elementary, we support getting parameters from both CLI command arguments and a config.yml file. To easily support that we introduced a new class called Config, this class loads the config.yml into a config dictionary. Then for each available arguement, we either try to use the provided argument or we try to load it from this config dictionary. [See this example here](https://github.com/elementary-data/elementary/blob/master/elementary/config/config.py#L92) to see how it’s implemented for the Slack parameters.
- [monitor/data_monitoring/data_monitoring_alerts](https://github.com/elementary-data/elementary/blob/master/elementary/monitor/data_monitoring/data_monitoring_alerts.py) -
    - This is the class responsible for the high-level alerts flow. The flow consists of sending all the different alert types (db test failures / warnings, elementary anomalies, model failures, source freshness issues) and then marking them as sent in the data warehouse.
    - Here we need to generalize [this](https://github.com/elementary-data/elementary/blob/master/elementary/monitor/data_monitoring/data_monitoring_alerts.py#L69) function to be called `_send_alerts_and_update_in_db` instead of `_send_alerts_to_slack`. This logic [here](https://www.notion.so/MS-teams-99c3c7de3b9048a9a76b0825b6b2a226) is the one responsible for sending a single alert to slack so we should extract it to a separate function called `_send_alert_to_slack`. Then we should add a similar function that is called `_send_alert_to_ms_teams`.  Now `_send_alerts_and_update_in_db` could decide if it should use `_send_alert_to_slack` or `_send_alert_to_ms_teams` based on what’s saved in `self.config`.
    - You’ll need to add two new main things to fully implement `_send_alert_to_ms_teams` -
        - MS teams clients (should be implemented under clients/ms_teams). These should be wrappers for the way Elementary communicates with ms teams. For slack we implemented two clients, one that is based off a [webhook](https://github.com/elementary-data/elementary/blob/master/elementary/clients/slack/client.py#L176) and one that is based off a [token](https://github.com/elementary-data/elementary/blob/master/elementary/clients/slack/client.py#L70). The reason why we have both is because webhook has a simpler setup but it doesn’t support uploading files in Slack. To implement sending the Elementary html report we added a client that is based on token. Feel free to start with what is more simpler for you in MS teams.
            - You can start by implementing only sending messages and not files as a start. We can always add this functionality later on. See the [SlackClient](https://github.com/elementary-data/elementary/blob/master/elementary/clients/slack/client.py#L19) interface to get inspiration for the clients interface.
        - Add to the Alert interface [here](https://github.com/elementary-data/elementary/blob/master/elementary/monitor/alerts/alert.py#L71) a new to_ms_teams interface function (please note, this should be just the interface and not the implementation itself). This should be the interface to the function that formats the alert to the ms teams message structure. Now for each and every alert object (TestAlert, DbtTestAlert, ElementaryTestAlert, SourceFreshnessAlert, ModelAlert, MalformedAlert) we should implement the to_ms_teams function. This function should format the alert attributes to the MS teams message structure. We should aim for a similar format of messages as the format that we have today on Slack. Happy to provide screenshots if needed.

**Comment by Januka-Peiris-Assured** [comment]
Thank you for that. Could you send me some screenshots of the sample messages please?

**Comment by christopherekfeldt** [comment]
Our company would love this feature :)

**Comment by iioo661** [comment]
Also interested in this feature.

**Comment by sharma-vaibhav1** [comment]
We would also like to see support for zoom client alerts along with slack and MS Team alerts. I think it will be a good idea to provide a generic webhook message that can be personalized as per the needs of the platforms

**Comment by DataGobes** [comment]
We need this as well, I implemented elementary today into our project and used slack but getting slack approved in our company is a potential problem. No contribution started on this yet? Then I might want to have a go at it with above instructions.

**Comment by ellakz** [comment]
Hello to everyone who offered to contribute this feature :)
We recently made some changes to decouple Slack from our alerts fetcher and formatter, so that it'll be easier to add new alert integrations!
Following those changes, I created an [integration contribution guide](https://github.com/elementary-data/elementary/blob/master/elementary/monitor/data_monitoring/alerts/integrations/README.md) that specifies exactly what needs to be done in order to fully support a new alert integration.
I believe supporting MS Teams for alerts only (via webhook) will take something like 2-5 days.
I also have a draft of the Teams message format that I can share, so that may also speed things up.
If this somehow encourages anyone on this thread to take this on we will be happy to assist!
